protocol/validation/validation.go

```diff
@@ -0,0 +1,466 @@
+	if err != nil {
+		return errors.Wrap(err, "checking value source")
+	}
+
+	var dest bc.ValueDestination
+	switch ref := vs.Entry.(type) {
+	case *bc.Issuance:
+		if vs.Position != 0 {
+			return errors.Wrapf(errPosition, "invalid position %d for issuance source", vs.Position)
+		}
+		dest = ref.Witness.Destination
+
+	case *bc.Spend:
+		if vs.Position != 0 {
+			return errors.Wrapf(errPosition, "invalid position %d for spend source", vs.Position)
+		}
+		dest = ref.Witness.Destination
+
+	case *bc.Mux:
+		if vs.Position >= uint64(len(ref.Witness.Destinations)) {
+			return errors.Wrapf(errPosition, "invalid position %d for %d-destination mux source", vs.Position, len(ref.Witness.Destinations))
+		}
+		dest = ref.Witness.Destinations[vs.Position]
+
+	default:
+		return errors.Wrapf(errEntryType, "value source is %T, should be issuance, spend, or mux", vs.Entry)
+	}
+
+	if dest.Ref != vstate.entryID {
+		return errors.Wrapf(errMismatchedReference, "value source for %x has disagreeing destination %x", vstate.entryID[:], dest.Ref[:])
+	}
+
+	if dest.Position != vstate.sourcePos {
+		return errors.Wrapf(errMismatchedPosition, "value source position %d disagrees with %d", dest.Position, vstate.sourcePos)
+	}
+
+	if dest.Value != vs.Value {
+		return errors.Wrapf(errMismatchedValue, "source value %v disagrees with %v", dest.Value, vs.Value)
+	}
+
+	return nil
+}
+
+func checkValidDest(vs *validationState, vd *bc.ValueDestination) error {
+	var src bc.ValueSource
+	switch ref := vd.Entry.(type) {
+	case *bc.Output:
+		if vd.Position != 0 {
+			return errors.Wrapf(errPosition, "invalid position %d for output destination", vd.Position)
+		}
+		src = ref.Body.Source
+
+	case *bc.Retirement:
+		if vd.Position != 0 {
+			return errors.Wrapf(errPosition, "invalid position %d for retirement destination", vd.Position)
+		}
+		src = ref.Body.Source
+
+	case *bc.Mux:
+		if vd.Position >= uint64(len(ref.Body.Sources)) {
+			return errors.Wrapf(errPosition, "invalid position %d for %d-source mux destination", vd.Position, len(ref.Body.Sources))
+		}
+		src = ref.Body.Sources[vd.Position]
+
+	default:
+		return errors.Wrapf(errEntryType, "value destination is %T, should be output, retirement, or mux", vd.Entry)
+	}
+
+	if src.Ref != vs.entryID {
+		return errors.Wrapf(errMismatchedReference, "value destination for %x has disagreeing source %x", vs.entryID[:], src.Ref[:])
+	}
+
+	if src.Position != vs.destPos {
+		return errors.Wrapf(errMismatchedPosition, "value destination position %d disagrees with %d", src.Position, vs.destPos)
+	}
+
+	if src.Value != vd.Value {
+		return errors.Wrapf(errMismatchedValue, "destination value %v disagrees with %v", src.Value, vd.Value)
+	}
+
+	return nil
+}
+
+// ValidateBlock validates a block and the transactions within.
+// The consensus program is executed only if prev is non-nil and
+// runProg is true.
+func ValidateBlock(b, prev *bc.BlockEntries, initialBlockID bc.Hash, runProg bool) error {
```

Comment:
Instead of using a boolean `runProg` flag, should we split into multiple functions? That way when you see a function's invocation, you can tell by it's name what it's doing. It's much harder to interpret a trailing `true` or `false`.